Parsing improvements

drewnoakes opened the pull request:
Changes proposed in this pull request:

- Add faster overload of `GetRemoteName` (used O(N) from `GitModule.GetTreeRefs`) which avoids additional git process execution to obtain remote names (`git remote show`)
- Add tests for `GitModule.GetTreeRefs`
- Use regex in `GitModule.GetTreeRefs` and reduce allocations
- Make `GitRef.Name` immutable
- Create `RefName` class for all utilities relating to ref names (i.e. `refs/remotes/origin/master` etc), moving several existing methods/constants to it, and adding unit tests

What did I do to test the code and ensure quality:
 - Added unit tests to modified code
 - Manual testing

Has been tested on:
 - GIT 2.16
 - Windows 10

RussKie commented on GitCommands/Git/RefName.cs:
```diff
@@ -0,0 +1,96 @@
+
+namespace GitCommands
+{
+    public static class RefName
```
For consistency I suggest we call it `GitRefName` or maybe even `GitRef`, since this class does not represent ref name, but provides capability to get it.

drewnoakes replied:
`GitRef` already exists, so went with `GitRefName`.

RussKie replied:
Aye, didn't see it. 👍

RussKie commented:
Could you please break the commit into several smaller ones with single purpose - e.g. git-ref as one, stash as the other etc.
It makes it easier to review. Thank you

drewnoakes commented:
> Could you please break the commit into several smaller ones with single purpose 

Created #4705 #4706 #4707. Couldn't split this further due to dependencies between commits. There's a natural progression here.

RussKie commented:
All the mentioned PRs have been merged.
Is this PR still relevant?

drewnoakes commented:
> All the mentioned PRs have been merged. Is this PR still relevant?

Yes, this PR still contains what was not split out into the merged PRs. When I split them out, I also updated this PR's primary comment/message to reflect what the PR aims to achieve.

I've rebased and fixed the conflict.

drewnoakes commented:
Rebased to fix conflicts and squashed.

RussKie commented:
Sorry Drew, but you will need to rebase again...

drewnoakes commented:
> Sorry Drew, but you will need to rebase again...

No worries. Done.
